### PR TITLE
fix(history): save status on version create

### PR DESCRIPTION
**Pull request by markkaylor**
### What does it do?

- Use the actual created entry as the data stored for the version instead of the context params
- Use the content-manager's `document-metadata` service to compute the status of the document version
- 
- Update the e2e tests to account for both publish and modified statuses

Unrelated
- Make the getLocaleDictionary private

### Why is it needed?

- To correctly save and display a history version's status

### Note

- unpublish actions are broken since the middleware does not have the data that was unpublished

### How to test it?

- Go to a content-type in draft state
- Make a change and save
- Publish and save
- Make another change and save
- Go to the history page. You should see something like modified, draft, published, draft

### CODE REVIEW

File: packages/core/admin/admin/src/components/FormInputs/Time.tsx [hunk 1]
```diff
@@ -32,7 +32,7 @@ const TimeInput = forwardRef<HTMLInputElement, InputProps>(
         }}
         onClear={() => field.onChange(name, undefined)}
         required={required}
-        value={field.value}
+        value={field.value ?? ''}
```
Comment (markkaylor) [inline packages/core/admin/admin/src/components/FormInputs/Time.tsx:35]: ℹ️ Otherwise I have an error coming from the DS trying to do regex string match on null. I saw DatePicker was doing this as well.

File: packages/core/admin/admin/src/content-manager/history/components/VersionContent.tsx [hunk 2]
```diff
@@ -43,6 +42,7 @@ const VersionContent = () => {
             if (panel.some((row) => row.some((field) => field.type === 'dynamiczone'))) {
               const [row] = panel;
               const [field] = row;
+
```
Comment (markkaylor) [inline packages/core/admin/admin/src/content-manager/history/components/VersionContent.tsx:45]: ℹ️ I didn't mean to make this change but I'm not too mad about it 🙃

File: packages/core/admin/admin/src/components/FormInputs/Time.tsx (hunk 1 above)
Comment (remidej) [reply packages/core/admin/admin/src/components/FormInputs/Time.tsx:35]: I don't mind but I can't find the place you're referring to

Comment (markkaylor) [reply packages/core/admin/admin/src/components/FormInputs/Time.tsx:35]: This is where the error comes from in the DS I believe:
https://github.com/strapi/design-system/blob/5062e288af88c4aae8cf36fd52e3600fa36044c3/packages/strapi-design-system/src/TimePicker/TimePicker.tsx#L14

ahh my bad it was StringInput that did this:
https://github.com/strapi/strapi/blob/f237ce8c3df66beb36d02a96297bfdd3c24c838f/packages/core/admin/admin/src/components/FormInputs/String.tsx#L31

File: packages/core/content-manager/server/src/history/services/history.ts [hunk 3]
```diff
@@ -68,23 +107,17 @@ const createHistoryService = ({ strapi }: { strapi: LoadedStrapi }) => {
...
-        const result = (await next(context)) as any;
-
         // Prevent creating a history version for an action that wasn't actually executed
         await strapi.db.transaction(async ({ onCommit }) => {
           onCommit(() => {
             this.createVersion({
               contentType: contentTypeUid,
-              relatedDocumentId: 'documentId' in result ? result.documentId : context.args[0],
               locale: params.locale,
               // TODO: check if drafts should should be "modified" once D&P is ready
```
Comment (remidej) [inline packages/core/content-manager/server/src/history/services/history.ts:116]: this comment can be removed

File: packages/core/content-manager/server/src/history/services/history.ts [hunk 4]
```diff
@@ -43,19 +78,23 @@ const createHistoryService = ({ strapi }: { strapi: LoadedStrapi }) => {
...
+        // @ts-expect-error ContentType is not typed correctly on the context
+        const contentTypeUid = context.contentType.uid;
         // Ignore content types not created by the user
         if (!contentTypeUid.startsWith('api::')) {
           return next(context);
         }
 
+        const params = context.args.at(-1) as any;
+        const result = (await next(context)) as any;
+        const data = context.action === 'publish' ? result.versions[0] : result;
```
Comment (remidej) [inline packages/core/content-manager/server/src/history/services/history.ts:94]: It would be useful to have a comment explaining the different cases

File: packages/core/content-manager/server/src/history/services/history.ts [hunk 5]
```diff
@@ -33,6 +34,40 @@ const createHistoryService = ({ strapi }: { strapi: LoadedStrapi }) => {
     return Math.min(licenseRetentionDays, DEFAULT_RETENTION_DAYS);
   };
 
+  // TODO: Refactor so i18n can interact with history without history itself being concerned about i18n
```
Comment (remidej) [inline packages/core/content-manager/server/src/history/services/history.ts:37]: Are we sure we want to do something about that todo in the end? @joshuaellis

File: packages/core/content-manager/server/src/history/services/history.ts [hunk 6]
```diff
@@ -33,6 +34,40 @@ const createHistoryService = ({ strapi }: { strapi: LoadedStrapi }) => {
...
+    data: HistoryVersions.CreateHistoryVersion['data']
+  ) => {
+    try {
+      const documentMetadataService = strapi.plugin('content-manager').service('document-metadata');
+      const meta = await documentMetadataService.getMetadata(contentTypeUid, data);
+
+      return documentMetadataService.getStatus(data, meta.availableStatus);
+    } catch (error) {
+      console.error(error);
+    }
```
Comment (remidej) [inline packages/core/content-manager/server/src/history/services/history.ts:68]: I'm curious to hear your thinking here to catch/silence the error

**Review by remidej — COMMENTED** [review]
not tested yet, but thanks for figuring out the middleware thing!

File: packages/core/content-manager/server/src/history/services/history.ts (hunk 6 above)
Comment (markkaylor) [reply packages/core/content-manager/server/src/history/services/history.ts:68]: I left a comment [here](https://github.com/strapi/strapi/pull/19737#discussion_r1521474444). I should test this again by throwing an error in the getMetadata function. I had a couple instances during development where I had an error tryin to get the status and it meant in the CM I couldn't create an entry. I don't think history should get in the way of creating content. Wdyt?

Comment (markkaylor) [reply packages/core/content-manager/server/src/history/services/history.ts:68]: Actually nvm seems to create the entry no matter what. I think I can remove this

Comment (remidej) [reply packages/core/content-manager/server/src/history/services/history.ts:68]: probably because it's called after awaiting next()

File: packages/core/content-manager/server/src/history/services/history.ts [hunk 7]
```diff
@@ -43,19 +79,35 @@ const createHistoryService = ({ strapi }: { strapi: LoadedStrapi }) => {
...
         if (!contentTypeUid.startsWith('api::')) {
           return next(context);
         }
 
+        const result = (await next(context)) as any;
+        /**
+         * Publish actions store the created data on the result's versions array.
+         * Otherwise, for create and update, the data is the result.
+         *
+         * TODO: Fix unpublish which is broken since the data that was unpublished is not passed to the middleware
```
Comment (remidej) [inline packages/core/content-manager/server/src/history/services/history.ts:98]: I'm fine with this for now. I just noticed that clicking "discard changes" in the UI doesn't create any history versions. Is that related, or fixable now?

**Review by remidej — COMMENTED** [review]
Code looks good, modified status and publishing too 👍

File: packages/core/content-manager/server/src/history/services/history.ts (hunk 7 above)
Comment (markkaylor) [reply packages/core/content-manager/server/src/history/services/history.ts:98]: Good thinking I haven't tried that yet, I'll take a quick look before merging so I can create a bug ticket for that and unpublish

Comment (markkaylor) [reply packages/core/content-manager/server/src/history/services/history.ts:98]: Sorry didn't read this part "or fixable now?", 🤣 I'll take a look

Comment (markkaylor) [reply packages/core/content-manager/server/src/history/services/history.ts:98]: I can fix that here :)

File: packages/core/content-manager/server/src/history/services/history.ts [hunk 8]
```diff
@@ -43,19 +79,38 @@ const createHistoryService = ({ strapi }: { strapi: LoadedStrapi }) => {
        * TODO: Fix the types for the middleware
        */
       strapi.documents.use(async (context, next) => {
-        // @ts-expect-error ContentType is not typed correctly on the context
-        const contentTypeUid = context.contentType.uid;
-        const params = context.args.at(-1) as any;
         // Ignore actions that don't mutate documents
-        if (!['create', 'update', 'publish', 'unpublish'].includes(context.action)) {
+        if (
+          !['create', 'update', 'publish', 'unpublish', 'discardDraft'].includes(context.action)
```
Comment (Marc-Roig) [inline packages/core/content-manager/server/src/history/services/history.ts:84]: you might want to cover the case where multiple locales are published/unpublished/discarded (using locale="*" param)

File: packages/core/content-manager/server/src/history/services/history.ts [hunk 9]
```diff
@@ -33,6 +40,35 @@ const createHistoryService = ({ strapi }: { strapi: LoadedStrapi }) => {
     return Math.min(licenseRetentionDays, DEFAULT_RETENTION_DAYS);
   };
 
+  const getLocaleDictionary = async () => {
+    if (!strapi.plugin('i18n')) {
+      return {};
+    }
```
Comment (Marc-Roig) [inline packages/core/content-manager/server/src/history/services/history.ts:46]: not a problem to add this , just that i18n will always be installed in v5

File: packages/core/content-manager/server/src/history/services/history.ts [hunk 10]
```diff
@@ -68,23 +123,16 @@ const createHistoryService = ({ strapi }: { strapi: LoadedStrapi }) => {
...
         await strapi.db.transaction(async ({ onCommit }) => {
           onCommit(() => {
             this.createVersion({
               contentType: contentTypeUid,
-              relatedDocumentId: 'documentId' in result ? result.documentId : context.args[0],
-              locale: params.locale,
-              // TODO: check if drafts should should be "modified" once D&P is ready
-              status: params.status,
-              data: omit(fieldsToIgnore, params.data),
+              locale: data.locale,
```
Comment (Marc-Roig) [inline packages/core/content-manager/server/src/history/services/history.ts:131]: this might be empty if user expect using the default locale of the app

File: packages/core/content-manager/server/src/history/services/history.ts [hunk 11]
```diff
@@ -43,19 +79,38 @@ const createHistoryService = ({ strapi }: { strapi: LoadedStrapi }) => {
...
+            : result;
+        /**
+         * The documentId should exist on the created data, fallback to context to handle the broken unpublish action
+         *
+         * TODO: Always get the documentId from the same place (data) once unpublish is fixed
+         */
+        const relatedDocumentId = data.documentId ?? context.args[0];
+        // Compute the status of the version
+        const status = await getVersionStatus(contentTypeUid, data);
         const fieldsToIgnore = [
```
Comment (Marc-Roig) [inline packages/core/content-manager/server/src/history/services/history.ts:114]: not sure of which status it is at the moment, but I believe localizations fields was (or is going to be )  removed

Comment (markkaylor) [reply packages/core/content-manager/server/src/history/services/history.ts:114]: ok sounds good thanks, I'll go ahead and remove it from the list here then

File: packages/core/content-manager/server/src/history/services/history.ts (hunk 8 above)
Comment (markkaylor) [reply packages/core/content-manager/server/src/history/services/history.ts:84]: ok, in that case would the versions array contain several documents?

Comment (markkaylor) [reply packages/core/content-manager/server/src/history/services/history.ts:84]: As it turns out we only need to support requests coming from the CM. Do you know if it is / will be possible to publish in several locales at once from the CM?

I think we will also need to handle this case for bulk publish/unpublish for a single locale as well but I prefer to set that up in another PR when bulk actions are working in v5.

Comment (markkaylor) [reply packages/core/content-manager/server/src/history/services/history.ts:84]: And we will need to filter out the requests that do not come from the content manager.

Comment (Marc-Roig) [reply packages/core/content-manager/server/src/history/services/history.ts:84]: >. Do you know if it is / will be possible to publish in several locales at once from the CM?
Yep! we will support that option

Comment (markkaylor) [reply packages/core/content-manager/server/src/history/services/history.ts:84]: Thanks!! I've created another ticket to handle the bulk action stuff in a follow up PR once all that's ready in the content manager.

Comment (Marc-Roig) [inline packages/core/content-manager/server/src/history/services/history.ts:84]: you might want to include clone too

**Comment by markkaylor** [comment]
Putting this back in review. I have no idea what's up with the e2e tests, unable to reproduce locally. Still looking for a fix.

Updates to test:
- Create a document and then duplicate it from the list view. The duplicated document should have a single version for its creation
-  History versions should only be created for changes made in the CM. Try to create a document somewhere using something like:
```ts
  await strapi.documents('api::something.something').create({ data: { test: 'plop' } });
```
it should create the document but not a history version.

File: packages/core/content-manager/server/src/history/services/history.ts [hunk 12]
```diff
@@ -43,48 +69,63 @@ const createHistoryService = ({ strapi }: { strapi: LoadedStrapi }) => {
...
+            ? // @ts-expect-error The context args are not typed correctly
+              { documentId: result.documentId, locale: context.args[0]?.locale }
+            : { documentId: context.args[0], locale: context.args[1]?.locale };
+
+        const locale = documentContext.locale ?? localesService.getDefaultLocale();
+        const document = await strapi
+          .documents(contentTypeUid)
+          .findOne(documentContext.documentId, {
+            locale,
+            populate: '*',
```
Comment (remidej) [inline packages/core/content-manager/server/src/history/services/history.ts:104]: Is there a way to populate only the ID everywhere? Because we'll never use the rest of the data, and it can take a lot of space for nothing

Comment (markkaylor) [reply packages/core/content-manager/server/src/history/services/history.ts:104]: oh yeah I'm sure I can just pick what we need here

Comment (markkaylor) [reply packages/core/content-manager/server/src/history/services/history.ts:104]: I take that back, I'm not sure I can populate all relation but only return the ID 🤔

Comment (remidej) [reply packages/core/content-manager/server/src/history/services/history.ts:104]: huh, that's too bad. And without populate I guess there's just no object entry at all?

Comment (remidej) [reply packages/core/content-manager/server/src/history/services/history.ts:104]: you could do some logic to trim the data before saving, but it's a bit annoying

Comment (remidej) [reply packages/core/content-manager/server/src/history/services/history.ts:104]: or you write a script that creates the populate object based on the schema (to identify where there are relations/images). Just as annoying, but faster

Comment (markkaylor) [reply packages/core/content-manager/server/src/history/services/history.ts:104]: I can actually remove the populate here. It is unrelated to handling the status 🙃 .

Comment (remidej) [reply packages/core/content-manager/server/src/history/services/history.ts:104]: but document ends up in the data column of the history version though right?

Comment (remidej) [reply packages/core/content-manager/server/src/history/services/history.ts:104]: if it is still needed, @Marc-Roig sent me this file which seems very handy: https://github.com/strapi/strapi/blob/f7d9269c38f7d4b8590c6f7b5c15981814c6d662/packages/core/core/src/services/document-service/utils/populate.ts

Comment (markkaylor) [reply packages/core/content-manager/server/src/history/services/history.ts:104]: - A populate object is sent along with the context so maybe we can use that. 
Screenshot 2024-03-14 at 16 13 58
- Even with populate removed here the statuses display correctly and the document ends up in the version's data column, it just doesn't include relations. Which I believe is more or less how it is now on v5/main.

Comment (remidej) [reply packages/core/content-manager/server/src/history/services/history.ts:104]: >the document ends up in the version's data column, it just doesn't include relations

yes but I think we need them though 🤔 It's like that currently on main because we never addressed relations, but I'm currently working on that, and I just need to know the IDs of related entries

Comment (remidej) [reply packages/core/content-manager/server/src/history/services/history.ts:104]: regarding the populate object, are we sure it will always be there? from what we've seen the context data is not very consistent, I'd prefer not to rely on it too much

and we don't need all the default fields (strapi_stage, strapi_assignee, created/updatedBy), only the ones that were created by the user

Comment (markkaylor) [reply packages/core/content-manager/server/src/history/services/history.ts:104]: > yes but I think we need them though 🤔 It's like that currently on main because we never addressed relations, but I'm currently working on that, and I just need to know the IDs of related entries

Right but should that be handled by this PR which is about "status"? In my opinion it feels out of scope and should be in a PR that is about relations. Let me know if I'm not understanding something because I'm confused why it should be done here. 

> regarding the populate object, are we sure it will always be there? from what we've seen the context data is not very consistent, I'd prefer not to rely on it too much

Yes I agree

Comment (remidej) [reply packages/core/content-manager/server/src/history/services/history.ts:104]: You're right I can address it in my PR, it's very related to what I'm doing now

**Review by remidej — APPROVED** [review]

**Comment by markkaylor** [comment]
The failing e2e tests are unrelated to the changes here and I am unable to reproduce locally

**Review by remidej — APPROVED** [review]
the locale thing is fixed 👍
